rust-src/concordium_base/src/web3id/v1/proofs.rs

```diff
@@ -0,0 +1,1157 @@
+    ConcordiumProofType, ConcordiumZKProof, ContextInformation, CredentialMetadataV1,
+    CredentialProofPrivateInputs, CredentialV1, CredentialVerificationMaterial,
+    IdentityBasedCredentialV1, IdentityBasedSubjectClaims, IdentityCredentialEphemeralId,
+    IdentityCredentialEphemeralIdDataRef, IdentityCredentialProofPrivateInputs,
+    IdentityCredentialProofs, IdentityCredentialSubject, IdentityCredentialVerificationMaterial,
+    PresentationV1, ProveError, RequestV1, SubjectClaims, VerifyError,
+};
+use rand::{CryptoRng, Rng};
+
+impl<P: Pairing, C: Curve<Scalar = P::ScalarField>, AttributeType: Attribute<C::Scalar>>
+    PresentationV1<P, C, AttributeType>
+{
+    /// Get an iterator over the metadata for each of the verifiable credentials
+    /// in the order they appear in the presentation. This contains data that need to be
+    /// verified externally and also data that is needed to look up [`CredentialVerificationMaterial`].
+    /// Hence, proper handling of the metadata is required for verifying the presentation. An implementation of handling
+    /// the metadata may be found in the [Rust SDK `web3id` module](https://docs.rs/concordium-rust-sdk/latest/concordium_rust_sdk/web3id/index.html)
+    pub fn metadata(&self) -> impl ExactSizeIterator<Item = CredentialMetadataV1> + '_ {
+        self.verifiable_credentials.iter().map(|cp| cp.metadata())
+    }
+
+    /// Verify a presentation and the contained credentials in the context of the provided verification material.
+    /// In case of success, returns the [`RequestV1`] which contains the claims the presentation
+    /// verifies.
+    ///
+    /// Notice: This only verifies the cryptographic consistency of the data.
+    /// It does not check metadata, such as expiry. This should be checked
+    /// separately by the verifier. See [`CredentialMetadataV1`].
+    pub fn verify<'a>(
+        &self,
+        global_context: &GlobalContext<C>,
+        verification_material: impl ExactSizeIterator<Item = &'a CredentialVerificationMaterial<P, C>>,
+    ) -> Result<RequestV1<C, AttributeType>, VerifyError> {
+        let mut transcript = RandomOracle::domain("ConcordiumVerifiablePresentationV1");
+        append_context(&mut transcript, &self.presentation_context);
+        transcript.append_message(b"ctx", &global_context);
+
+        let mut request = RequestV1 {
+            challenge: self.presentation_context.clone(),
+            subject_claims: Vec::new(),
+        };
+
+        if verification_material.len() != self.verifiable_credentials.len() {
+            return Err(VerifyError::VeficationMaterialMismatch);
+        }
+
+        for (i, (verification_material, credential)) in verification_material
+            .zip(&self.verifiable_credentials)
+            .enumerate()
+        {
+            request.subject_claims.push(credential.claims());
+
+            if !credential.verify(global_context, &mut transcript, verification_material) {
```

Comment:
Note: This ties the credentials proofs together, e.g. the order of the credentials must match. 
This is not necessarily required, but it should not hurt us